sound/soc/sof/intel/hda.c

```diff
@@ -72,22 +72,38 @@ int hda_ctrl_dai_widget_setup(struct snd_soc_dapm_widget *w, bool setup)
 	if (setup) {
 		ret = sof_widget_setup(sdev, swidget);
 		if (ret < 0) {
-			dev_err(sdev->dev, "error: setting up DAI widget %s\n", w->name);
+			dev_err(sdev->dev, "error: failed setting up DAI widget %s\n", w->name);
 			return ret;
 		}
 
 		/* send DAI_CONFIG IPC */
-		return sof_ipc_tx_message(sdev->ipc, config->hdr.cmd, config, config->hdr.size,
+		ret = sof_ipc_tx_message(sdev->ipc, config->hdr.cmd, config, config->hdr.size,
 					  &reply, sizeof(reply));
+		if (ret < 0) {
+			dev_err(sdev->dev, "error: failed setting DAI config for %s\n", w->name);
+			return ret;
+		}
+
+		sof_dai->configured = true;
+
+		return 0;
 	}
```

Comment:
this was actually the case before this patch but now it's becoming even more apparent. This function actually contains two functions. It's one of those cases:
```
void x(bool mode,...)
{
	if (mode) {
		do_one_thing(...);
		return;
	}
	do_another_thing(...);
}
```
which isn't a very good approach IMHO. Maybe just make two functions out of it: one for `setup == true` and another one for `false`.

Reply:
this will mean that a few linesof code will be duplicated. If that is the consensus, I will split it up. @plbossart @kv2019i what do you think?

Reply:
I agree it's a bit odd, I prefer 

````
if (mode)
   do_one_thing
common_part
if (!mode)
   do_another_thing
````

Reply:
ok i've split up the function into 2. Hopefully a bit more intuitive now.